presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -87,150 +92,189 @@ private Rewriter(PlanNodeIdAllocator idAllocator, Metadata metadata, Session ses
-                return Optional.empty();
-            }
+            PlanNode source = context.rewrite(node.getSource());
+
+            TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+            // change the source node to RowNumber or TopNRowNumber and leave the FilterNode around for simplicity.
+            // for some simple filters (rn < 5), the filter might be redundant, but in most cases, it will not impact the query performance too much
+            if (source instanceof RowNumberNode) {
+                Symbol rowNumberSymbol = ((RowNumberNode) source).getRowNumberSymbol();
+                Optional<Integer> upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
 
-            // try to get the upper bound of row number
-            Optional<Long> upperBound = Optional.empty();
-            Range span = rowNumberRangeSet.getSpan();
-
-            if (!span.getHigh().isUpperUnbounded()) {
-                switch (span.getHigh().getBound()) {
-                    case EXACTLY:
-                        upperBound = Optional.of((Long) span.getHigh().getValue());
-                        break;
-                    case BELOW:
-                        upperBound = Optional.of((Long) span.getHigh().getValue() - 1);
-                        break;
-                    default:
-                        break;
+                if (upperBound.isPresent()) {
+                    source = mergeLimit(((RowNumberNode) source), upperBound.get());
+                    Optional<PlanNode> filterNode = transformFilterNode(node, source, rowNumberSymbol, upperBound.get());
+                    return filterNode.orElse(source);
                 }
             }
+            else if (source instanceof WindowNode && canOptimizeWindowFunction((WindowNode) source)) {
+                WindowNode windowNode = (WindowNode) source;
+                Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+                Optional<Integer> upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
 
-            if (upperBound.isPresent() && upperBound.get() <= Integer.MAX_VALUE) {
-                return Optional.of(upperBound.get().intValue());
+                if (upperBound.isPresent()) {
+                    source = convertToTopNRowNumber(windowNode, upperBound.get());
+                    Optional<PlanNode> filterNode = transformFilterNode(node, source, rowNumberSymbol, upperBound.get());
+                    return filterNode.orElse(source);
+                }
             }
-            return Optional.empty();
+
+            return context.replaceChildren(node, ImmutableList.of(source));
         }
 
-        private static boolean canOptimizeWindowFunction(WindowNode node)
+        private Optional<PlanNode> transformFilterNode(FilterNode filterNode, PlanNode source, Symbol rowNumberSymbol, int upperBound)
```

Comment:
you can technically just call this rewriteFilterSource() and have it return the node directly, rather than an optional that is OR'ed with the original source.